Added List.sort(comp) to List module

mathewmariani opened the pull request:
Added `sort(comp)` function to List using an in-place quicksort implementation.

CohenArthur commented:
Your doc example shows it, but it would be nice to have some documentation on the expected behavior of `comp()`. For example, saying it must take two arguments and return a bool, etc etc

mathewmariani commented:
> Your doc example shows it, but it would be nice to have some documentation on the expected behavior of `comp()`. For example, saying it must take two arguments and return a bool, etc etc

Thanks for pointing that out. I just updated the docs and added a default `comp` functions

mhermier commented:
From my experience at using the language, I would recommend to change `sort()` to be a uni-line. Stub functions are only meant to provide default values or simply computed values, they should behave exactly like their parent and also return the same thing. While technically equivalent in your case, making it already there, express intent more, avoid possible future bugs and late fixing.

So as a rule of thumb; if you write a one liner, it is highly likely you want to return the computed value, so use the real one liner expression (that implicitly returns the computed value). I know, it is not really intuitive or does not scale producing changes, but the language is like this...

mathewmariani commented:
> From my experience at using the language, I would recommend to change `sort()` to be a uni-line. Stub functions are only meant to provide default values or simply computed values, they should behave exactly like their parent and also return the same thing. While technically equivalent in your case, making it already there, express intent more, avoid possible future bugs and late fixing.
> 
> So as a rule of thumb; if you write a one liner, it is highly likely you want to return the computed value, so use the real one liner expression (that implicitly returns the computed value). I know, it is not really intuitive or does not scale producing changes, but the language is like this...

Just to clarify you mean a uni-line (stue function) like so `sort() { sort {|a, b| a < b} }` or is there a Wren style that im missing?

mhermier commented:
Yes this one, it act as a stub with implicit returning.

bjorn commented on test/core/list/sort.wren:
```diff
@@ -0,0 +1,10 @@
+var list = [4, 1, 3, 2].sort()
+System.print([4, 1, 3, 2].sort()) // expect: [1, 2, 3, 4]
```
Does this test pass? I don't see `List.sort()` returning any value, so I would expect this to print `null`. Did you mean to `return this` at the end of `sort`?

mathewmariani replied:
With this [issue](https://github.com/wren-lang/wren/issues/801) open the tests don't pass.

mhermier replied:
@bjorn one liner body/method/function does implicitly return.

mathewmariani replied:
@mhermier the tests were returning null

bjorn replied:
@mhermier I know, but since the `sort(_)` function was returning `null`, so did the one-line version of `sort()`.

mathewmariani replied:
So `sort(_)` should return `this` and the one-liner `sort()` should be reverted back to a stub function?

bjorn commented on src/vm/wren_core.wren:
```diff
@@ -323,6 +323,43 @@ class List is Sequence {
 
+  sort() {
+    sort {|a, b| a < b }
+    return this
```
I think it's rather weird for `sort()` to return `this` but not `sort(comparer)`. I'd suggest changing this back to the one-line version and putting the `return this` in `sort(comparer)`.

mhermier commented:
Does the tests pass now ?

mathewmariani commented:
> Does the tests pass now ?

Not this the [stack corruption](https://github.com/wren-lang/wren/issues/761) issue

mhermier commented:
I think the test should be a bit more atomic to be more interesting.

mathewmariani commented:
I rebased from `main` and can confirm the tests are passing

bjorn commented:
@mathewmariani It seems you did a rebase followed by a merge. Please get rid of the merge for force-push just the rebased version of your branch.

mathewmariani commented:
> @mathewmariani It seems you did a rebase followed by a merge. Please get rid of the merge for force-push just the rebased version of your branch.

I believe that's it

mathewmariani commented:
Should I resolve all the comments, or leave that up to the commenters?

bjorn commented:
@mathewmariani I see most of my comments marked as "Outdated" and I see no way to resolve them (neither on ones that haven't been marked outdated), so I guess that's on you.

mathewmariani commented:
Anything else needed for this to be merged?

bjorn commented on projects/xcode/wren_test.xcodeproj/project.xcworkspace/contents.xcworkspacedata:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Workspace
```
You resolved this comment but the file is still being added by the pull request.

mathewmariani replied:
@bjorn should be done now

ChayimFriedman2 commented:
Nice work. Three things, however:

 1. I think it's preferred to implement this in C. Wren prefers to implement the core in C (speed).
 2. You implemented a quick sort. Do we need a stable sort instead?
 3. Should we add a non in-place sort, too? (Maybe a method in `Sequence`). It can be just:
    ```wren
    class Sequence {
      // ...
      sort() { sort {|a, b| a < b } }
      sort(comparer) { toList.sort(comparer) }
    }
    ```

avivbeeri commented:
I don't think it's semantically appropriate to have a Sequence.sort method, because a sequence could in theory be infinite in length (and they can be lazily evaluated), so a sorting operation would be impossible or intractable.

ruby0x1 commented:
Also a sort function in c runs into reentrancy (specifically, it's not possible when a user supplied compare function is involved)

ChayimFriedman2 commented:
@ruby0x1 I think the reentrancy is about to be solved, am I wrong?

@avivbeeri I know that, be that still holds in Python for example, and it does have a `sorted()` function that operates on iterables.

mathewmariani commented:
@ChayimFriedman2 I implemented Quick Sort since it's the algorithm used in Lua's `table.sort`
@ruby0x1 is there anything left that I should be adding for this MR?

ruby0x1 commented:
I'll look closely soon, busy with some other Wren stuff locally.

ruby0x1 commented on src/vm/wren_core.wren:
```diff
@@ -323,6 +323,41 @@ class List is Sequence {
+  }
+
+  quicksort_(a, b, comparer) {
+    if (a < b) {
```
@mathewmariani this is using `<` despite comparer being provided. 
I would expect `if(comparer.call(a, b)) {` to be the condition instead?

bjorn replied:
Actually `a` and `b` are the start and end indexes of the range that's being sorted. Since the range is inclusive, the comparison is checking whether the range has more than 1 element.

Maybe there could be more intuitive names for these variables.

ruby0x1 replied:
Oh right, I misread there. I agree with better naming.

mathewmariani replied:
I was thinking `i` and `j` since that's what most algorithms & data-structure books might use. But, `start` and `end` can be used as well.

ruby0x1 commented on doc/site/modules/core/list.markdown:
```diff
@@ -94,6 +94,25 @@ System.print(["a", "b", "c"].removeAt(1)) //> b
+System.print(list) //> [1, 2, 3, 4]
+</pre>
+
+A comparison function `comparer` can be provided to customise the element sorting. The comparison function must return a boolean value specifying the order in which elements should appear in the list. The default behaviour uses a `<` comparison.
```
This doesn't explain the boolean result, or the nature of the compare function :) We'd ideally do both, I can adjust the documentation after merge though, just wanted to give feedback.

> The comparison function accepts two arguments, which are the two values to compare, and must return a boolean value to indicate if the first value is smaller than the second value. If the function returns true, the first argument will be before the second in the sorted results. 
> 
> A compare function like `{|value, other| true }` will always put `value` before `other`. The default compare function is `{|a, b| a < b }`.

ruby0x1 commented on src/vm/wren_core.wren:
```diff
@@ -323,6 +323,41 @@ class List is Sequence {
     return other
   }
 
+  sort() { sort {|a, b| a < b } }
```
I suspect this might want to be `<=` for stable sort results, otherwise equal items will rearrange themselves on successive sorting, and won't be deterministic.

ChayimFriedman2 replied:
Quick sort isn't stable anyway.

ChayimFriedman2 replied:
https://stackoverflow.com/questions/13498213/quicksort-algorithm-stability

avivbeeri replied:
It can be implemented in a manner that makes it stable, at the cost of memory consumption: https://www.geeksforgeeks.org/stable-quicksort/

As an alternative, we could use a merge sort, which has similar properties, plus stability. However, it would also have some memory consumption requirements.

ChayimFriedman2 replied:
I know that, but unless we use a stable sort (which I suggested earlier), it's useless to discuss about `<` vs. `<=`.

mathewmariani replied:
Is there a specific reason we want a stable sort? My idea for using a non-stable quick-sort was that it's highly memory efficient which is perfect for, in my opinion, virtual memory environments like the WrenVM. It might have a worst-case scenario of running at O(n^2) but on average it should run O(n logn).

If a more efficient algorithm is needed it can be left to the user.

avivbeeri replied:
I think we can get away with it being unstable, so long as we document it clearly. From an embeddability point of view, the memory usage argument is quite compelling, in my opinion.

mathewmariani replied:
@avivbeeri Thank you! and yes, I will have to redo the documentation

mhermier commented:
To help comprehension, could you replace 'a' and 'b' with something like
'i' and 'j' where iterators are expected.

ChayimFriedman2 commented:
I think `start` and `end` would be better.

mathewmariani commented:
@ChayimFriedman2 @mhermier I'm leaning towards `i` and `j` since that's what you find in most books. I also like more compact code.

ruby0x1 commented:
@mathewmariani code is for humans, not machines, legibility > compactness. especially for wren, which strives to help people learn and understand, making the code clearer is part of that goal. In my code I used `low` `high` but `start` and `end` is also clear enough. I would prefer not to use single letter variables here.

ruby0x1 commented:
Thanks @mathewmariani , I'll probably tweak some things after merging, but looks good from here.
